### DISCUSSION TITLE
[help] How can I integrate `latexindent ` into alpheleia?

### DISCUSSION THREAD

**Discussion by avatar-lavventura** [radian-software/apheleia#78] [Q&A]
I have alias for latexindent: `alias latexindent="~/tools/latexindent.pl/latexindent.pl -l=~/indentconfig.yaml"`

I am able to run it as:

```bash
~/tools/latexindent.pl/latexindent.pl -l=~/indentconfig.yaml myfile.tex > output.tex
```

----------

But inside emacs I am having following error related to latexindent:

`*apheleia-latexindent-log*`

```
$ latexindent --logfile\=/dev/null
Can't open /dev/null at /usr/local/texlive/2021/texmf-dist/scripts/latexindent/LatexIndent/Document.pm line 165, <> line 328.
Command failed with exit code 2.
Fri Jan 14 12:41:34 2022 :: /Users/user/folder
$ latexindent --logfile\=/dev/null
Can't open /dev/null at /usr/local/texlive/2021/texmf-dist/scripts/latexindent/LatexIndent/Document.pm line 165, <> line 328.
Command failed with exit code 2.
```

How can I fix this error in order to run `latexindent` with my given config file?

**Comment by raxod502** [comment]
Is this on Windows or something? I'm confused why you'd be getting `Can't open /dev/null`.

But in any case, shell aliases only exist within the shell. They are not picked up by Emacs or by any other program that runs commands on your behalf. If you want to change the options that Apheleia passes to `latexindent`, you either need to edit `apheleia-formatters` or you need to create a wrapper script for `latexindent` that you put on your PATH.

**Comment by avatar-lavventura** [comment]
@raxod502 Its on macOS. // It would be nightmare to use emacs Windows :-)

Seems like `PATH` sees it :

```
$ which latexindent
/Library/TeX/texbin/latexindent
```

-----

seems like following wrapper worked but I was not able to apply `-l=$HOME/indentconfig.yaml` 

`latexindent` :
```
#!/bin/bash
~/tools/latexindent.pl/latexindent.pl $1
```

**Reply by raxod502** [reply]
You want `"$@"` instead of `$1` in your wrapper script, which will allow you to pass multiple arguments, and will also not break on whitespace. When you say you were "not able to apply" `-l=$HOME/indentconfig.yaml`, what do you mean? You added it to the wrapper script, and then when invoking the wrapper script, you got an error?

**Reply by avatar-lavventura** [reply]
When I use `-l=$HOME/indentconfig.yaml` it was not able to work properly in within the emacs, let me try with "$@"

**Reply by avatar-lavventura** [reply]
`latexindent.pl -l=$HOME/personalize/bin/localSettings.yaml "$@"` prints right away:

```
defaultIndent: "    "
verbatimEnvironments:
longtable: 1
tabularx: 1
specialBeginEnd:
If:
begin: '\\If'
    middle: '\\ElsIf'
    end: '\\EndIf'
lookForThis: 1
specialBeforeCommand: 1
```

So I had to run just: `latexindent.pl -l "$1"`

**Reply by raxod502** [reply]
Define "prints right away". How are you invoking this script, in what environment, and with what arguments, and where are you seeing the output?

**Comment by avatar-lavventura** [comment]
*Right away*: After the command is executed. 

`latexindent.pl -l=$HOME/personalize/bin/localSettings.yaml "$@"` instead of asking `Please enter text to be indented:` it prints the yaml file, which cause error in `emacs`.

```bash
$ latexindent.pl -l "$1"
Please enter text to be indented: (press CTRL+D when finished)

$ latexindent.pl -l=$HOME/personalize/bin/localSettings.yaml "$@"
defaultIndent: "    "
verbatimEnvironments:
longtable: 1
tabularx: 1
specialBeginEnd:
If:
begin: '\\If'
    middle:
    - '\\ElsIf'
    - '\\Else'
    end: '\\EndIf'
specialBeforeCommand: 1
```

environment: macOS Monterey Version 12.1 when I run with the `-l=` flag.

**Reply by raxod502** [reply]
You can't use `$1` and `$@` at the command line, they have to be from within a script, since they refer to the arguments that the script received. What would you expect `$@` to expand to in your second example? You haven't given any filename to substitute in.

I can't explain why your version of latexindent is parsing the argument to `-l` as a file to be indented, but that would be something to take up with the latexindent maintainers, I don't have any special insight there.

**Reply by avatar-lavventura** [reply]
Currently I have: 

```bash
$ latexindent --version
3.14, 2022-01-08
$ latexindent.pl -l "$1"  # works
$ latexindent.pl -l=$HOME/localSettings.yaml "$1"  # dose not work
```

I was able to resolve it by creating `.localSettings.yaml` forcing it to read from file in its home directory.

`$1` or `$@` should be a filename like `filename.tex` .

**Reply by raxod502** [reply]
Okay... sounds like you got it sorted, then?

**Reply by avatar-lavventura** [reply]
Yes sir with an ugly hack :)

**Reply by lispandfound** [reply]
Hey, I found this problem after two years, and I have solved it in a nicer way. The cause of the problem is that latexindent decides the yaml file it sees in the argument list is the file to format. So it dutifully "formats" the yaml settings and prints that back out again. To fix that, you should provide latexindent with `-` as an argument so it reads from stdin. I was able to get this working with the following entry in `apheleia-formatters`

```
(latexindent-conf "latexindent" "-m" "-l" ".indentconfig.yaml" "--logfile=/dev/null" "-")
```

This loads `.indentconfig.yaml` (in the same directory as the tex file) for local settings to apply if it exists, but reads the tex file from stdin like apheleia expects. Hope this helps anyone else who stumbles across the same problem as me!

**Reply by avatar-lavventura** [reply]
@lispandfound  I am having following error, it is normal:

```
Debugger entered--Lisp error: (void-function latexindent-conf)
  (latexindent-conf "latexindent" "-m" "-l" ".indentconfig.yaml" "--logfile=/dev/null" "-")
  eval-buffer(#<buffer  *load*> nil "/Users/alper/.emacs" nil t)  ; Reading at buffer position 112600
  load-with-code-conversion("/Users/alper/.emacs" "/Users/alper/.emacs" t t)
  load("~/.emacs" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode -0x14938f1d22b2b8b>) #f(compiled-function
() #<bytecode -0x1f3c686ddc0c1175>) t)
  command-line()
  normal-top-level()
  ```

**Reply by lispandfound** [reply]
Ah yeah, I should've been clearer. You'll want to add something like the following into your config file.

```emacs-lisp
(require 'apheleia) ;; Or do the following within your use-package declaration.

(add-to-list 'apheleia-formatters '(latexindent-conf "latexindent" "-m" "-l" ".indentconfig.yaml" "--logfile=/dev/null" "-"))

(setf (alist-get 'latex-mode apheleia-mode-alist) 'latexindent-conf)
(setf (alist-get 'LaTeX-mode apheleia-mode-alist) 'latexindent-conf)
;; etc. see `apheleia-mode-alist' for a complete list of the tex related modes you might want to set.
```

I ran this after installing apheleia in a fresh emacs with `emacs -Q` so this is does not depend on the specifics of my configuration. Then if you have a directory containing both a latex file and a config file like so

```bash
$ ls
.indentconfig.yaml
test.tex
``` 

The formatting options specified in .indentconfig.yaml will be applied when formatting test.tex.
